src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs

```diff
@@ -0,0 +1,368 @@
+    /// </summary>
+    /// <remarks>This is unsealed only to facilitate testing.</remarks>
+    public class PluginManager : IPluginManager, IDisposable
+    {
+        private static readonly Lazy<IPluginManager> Lazy = new Lazy<IPluginManager>(() => new PluginManager());
+        public static IPluginManager Instance => Lazy.Value;
+
+        private const string _idleTimeoutEnvironmentVariable = "NUGET_PLUGIN_IDLE_TIMEOUT_IN_SECONDS";
+        private const string _pluginPathsEnvironmentVariable = "NUGET_PLUGIN_PATHS";
+
+        private ConnectionOptions _connectionOptions;
+        private Lazy<IPluginDiscoverer> _discoverer;
+        private bool _isDisposed;
+        private IPluginFactory _pluginFactory;
+        private ConcurrentDictionary<PluginRequestKey, Lazy<Task<IReadOnlyList<OperationClaim>>>> _pluginOperationClaims;
+        private ConcurrentDictionary<string, Lazy<IPluginMulticlientUtilities>> _pluginUtilities;
+        private string _rawPluginPaths;
+
+        private static Lazy<int> _currentProcessId = new Lazy<int>(GetCurrentProcessId);
+
+        /// <summary>
+        /// Gets an environment variable reader.
+        /// </summary>
+        /// <remarks>This is non-private only to facilitate testing.</remarks>
+        public IEnvironmentVariableReader EnvironmentVariableReader { get; private set; }
+
+        private PluginManager()
+        {
+            Reinitialize(
+                new EnvironmentVariableWrapper(),
+                new Lazy<IPluginDiscoverer>(InitializeDiscoverer),
+                (TimeSpan idleTimeout) => new PluginFactory(idleTimeout));
+        }
+
+        /// <summary>
+        /// Creates a new plugin manager
+        /// </summary>
+        /// <remarks>This is public to facilitate unit testing. This should not be called from product code</remarks>
+        /// <param name="reader">An environment variable reader.</param>
+        /// <param name="pluginDiscoverer">A lazy plugin discoverer.</param>
+        /// <param name="pluginFactoryCreator">A plugin factory creator.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="reader" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="pluginDiscoverer" />
+        /// is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="pluginFactoryCreator" />
+        /// is <c>null</c>.</exception>
+        public PluginManager(IEnvironmentVariableReader reader,
+            Lazy<IPluginDiscoverer> pluginDiscoverer,
+            Func<TimeSpan, IPluginFactory> pluginFactoryCreator)
+        {
+            Reinitialize(
+                reader,
+                pluginDiscoverer,
+                pluginFactoryCreator);
+        }
+
+        /// <summary>
+        /// Disposes of this instance.
+        /// </summary>
+        public void Dispose()
+        {
+            if (!_isDisposed)
```

Comment:
isn't this singleton and lives as long as parent process lives? what cases it can fault into multiple calls into Dsipose?

Reply:
That's correct. 

I've added a comment to explain this. 

Basically since the PluginManager owns objects which implement IDisposable and should be implementing it, we have to provide the ability here. 

However, dispose is not part of the IPluginManager interface so therefore no user should be able to dispose it.